src/mission_companion.cpp

```diff
@@ -1563,3 +3685,672 @@ std::vector<item*> talk_function::loot_building(const tripoint site)
+    if( id != "FACTION_CAMP" ){
+        return cur_title;
+    }
+    const point omt_pos = ms_to_omt_copy( g->m.getabs( p.posx(), p.posy() ) );
+    point change;
+    switch (cur_tab){
+        case TAB_MAIN:
+            return _("Base Missions");
+        case TAB_N:
+            change.y--;
+            break;
+        case TAB_NE:
+            change.y--;
+            change.x++;
+            break;
+        case TAB_E:
+            change.x++;
+            break;
+        case TAB_SE:
+            change.y++;
+            change.x++;
+            break;
+        case TAB_S:
+            change.y++;
+            break;
+        case TAB_SW:
+            change.y++;
+            change.x--;
+            break;
+        case TAB_W:
+            change.x--;
+            break;
+        case TAB_NW:
+            change.y--;
+            change.x--;
+            break;
+    }
+    oter_id &omt_ref = overmap_buffer.ter( omt_pos.x + change.x, omt_pos.y + change.y, p.posz() );
+    std::string om_cur = omt_ref.id().c_str();
+    if( om_cur.find("faction_base_farm") != std::string::npos ){
+        return _("Farm Expansion");
+    }
+    if( om_cur.find("faction_base_garage_") != std::string::npos ){
+        return _("Garage Expansion");
+    }
+    if( om_cur.find("faction_base_kitchen_") != std::string::npos ){
+        return _("Kitchen Expansion");
+    }
+    if( om_cur == "field" || om_cur == "forest" || om_cur == "swamp" ){
+        return _("Empty Expansion");
+    }
+    return _("Base Missions");
+}
+
+std::map<std::string,std::string> talk_function::camp_recipe_deck( std::string om_cur ){
+    std::map<std::string,std::string> cooking_recipes;
+    if( om_min_level( "faction_base_kitchen_1", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_0.begin(), COOK_SKILL_0.end());
+    }
+    if( om_min_level( "faction_base_kitchen_2", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_1.begin(), COOK_SKILL_1.end());
+    }
+    if( om_min_level( "faction_base_kitchen_3", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_2.begin(), COOK_SKILL_2.end());
+    }
+    if( om_min_level( "faction_base_kitchen_4", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_3.begin(), COOK_SKILL_3.end());
+    }
+    if( om_min_level( "faction_base_kitchen_5", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_4.begin(), COOK_SKILL_4.end());
+    }
+    if( om_min_level( "faction_base_kitchen_6", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_5.begin(), COOK_SKILL_5.end());
+    }
+    if( om_min_level( "faction_base_kitchen_7", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_6.begin(), COOK_SKILL_6.end());
+    }
+    if( om_min_level( "faction_base_kitchen_8", om_cur ) || om_cur == "ALL" ){
+        cooking_recipes.insert(COOK_SKILL_7.begin(), COOK_SKILL_7.end());
+    }
```

Comment:
With the `array` option I described above, you could change it to a much cleaner:
```
size_t get_kitchen_level( const std::string &om_cur ) {
    static const std::map<std::string, size_t> kitchen_level = {{
        { "faction_base_kitchen_1", 1 },
        { "faction_base_kitchen_2", 2 },
        ...
    }};
    const auto iter = kitchen_level.find( om_cur );
    if( iter != kitchen_level.end() ) {
        return iter->second;
    }
    return 0;
}
std::vector<npc_cooking_recipe> talk_function::camp_recipe_deck( const std::string &om_cur ) {
    size_t kitchen_level = get_kitchen_level( om_cur );
    std::vector<npc_cooking_recipe> recipes;
    for( size_t i = 0; i < kitchen_level; i++ ) {
        recipes.insert( npc_recipes_at_skill_level[i].begin(), npc_recipes_at_skill_level[i].end() );
    }
    return recipes;
}
```

Then if you even needed to change things, you would only change them in one place instead of all the copied+pasted entries.

Reply:
Looking at adding the smith recipes now, is it still more practical to convert to a 
std::array<std::vector<npc_cook_recipe>, 8> recipes_at_skill;
system?  Add another dimension for each building (cook, camp, farm, smith ) that uses recipes?

I'm just using maps because they are super easy to use and auto-sort themselves, I have no idea if they are really shitty for run speed or memory.

Reply:
Performance is not a concern here, as long as it isn't checked every turn or anything like that.
It would be fine if it was a huge vector of structs containing skill, skill level, and recipe id, and you went through all existing recipes.

The real problem here is readability and maintainability. When you have 8 separate cases that only differ by one string and variable names. This is a sure sign that you could do better.